skimage/morphology/rolling_ball.py

```diff
@@ -0,0 +1,148 @@
+    if not isinstance(radius, float):
+        try:
+            radius = float(radius)
+        except ValueError:
+            raise ValueError(f"Radius should be float, was {type(radius)}")
+
+    if radius <= 0:
+        raise ValueError(f"Radius must be greater zeros, was {radius}")
+
+    try:
+        white_background = bool(white_background)
+    except ValueError:
+        raise ValueError(
+            f"white_background must be convertible to a boolean, "
+            f"was {type(white_background)}"
+        )
+
+    if not isinstance(input_img, np.ndarray):
+        raise ValueError(
+            f"input_img must be a np.ndarray, "
+            f"was {type(input_img)}"
+        )
+
+    if not issubclass(input_img.dtype.type, np.integer):
+        raise ValueError("Currently only integer images are supported.")
+
+    img = input_img.copy()
+    if white_background:
+        img = invert(img)
+
+    # sagitta assuming the position is where the ball touches
+    # the image umbra
+    L = np.arange(-radius, radius + 1)
+    X, Y = np.meshgrid(L, L)
+    distance = np.sqrt(X ** 2 + Y ** 2)
+    sagitta = radius - np.sqrt(radius ** 2 - distance ** 2)
+    sagitta[np.isnan(sagitta)] = np.Inf
+
+    kernel = np.array((X ** 2 + Y ** 2) <= radius ** 2, dtype=float)
+    kernel[kernel == 0] = np.Inf
+
+    kernel_size = np.array(kernel.shape)
+    img_original = img.copy()
+    img_size = img.shape
+    x = img
+    stride = (1, 1)
+
+    # pad the image
+    padding_amount = kernel_size // 2 * 2
+    half_pad = padding_amount // 2
+    img = np.Inf * np.ones(x.shape + padding_amount, dtype=x.dtype)
+    img[half_pad[0]:-half_pad[0], half_pad[1]:-half_pad[1]] = x
+    large_img_size = img.shape
+
+    # indices of kernel in image coords
+    x_idx = np.arange(kernel_size[1])
+    y_idx = large_img_size[1] * np.arange(kernel_size[0])
+    kernel_idx = (x_idx[np.newaxis, :] + y_idx[:, np.newaxis]).flatten()
+
+    # indices corresponding to each ancor of the kernel
+    # (top left corner instead of center)
+    x_ancors = np.arange(
+        large_img_size[1] - kernel_size[1] + 1, step=stride[1])
+    y_ancors = large_img_size[1] * \
+        np.arange(large_img_size[0] - kernel_size[0] + 1, step=stride[0])
+    ancor_offsets = (x_ancors[np.newaxis, :] +
+                     y_ancors[:, np.newaxis]).flatten()
+
+    # large images or kernel sizes don't fit into memory
+    # do it in batches instead
+    background = np.zeros(img_original.size)
+    batch_size = int(2 ** 10)
+    flat_img = img.flatten()
+    flat_sagitta = sagitta.flatten()
+    for low in range(0, len(ancor_offsets), batch_size):
+        high = np.minimum(low + batch_size, len(ancor_offsets))
+        filter_idx = ancor_offsets[low:high,
+                                   np.newaxis] + kernel_idx[np.newaxis, :]
+        background_partial = np.min(
+            flat_img[filter_idx] + flat_sagitta[np.newaxis, :], axis=1)
+        background[low:high] = background_partial
```

Comment:
Again, here, can you explain to me in english what we are iterating over, and what we are doing at each iteration? I think there's probably a fast algorithm in Cython lurking in here, but I don't see it yet! 😅

Reply:
I slide a window of kernel.size over the image (positions defined by ancor_offsets), processing `batch_size` windows each time. For each window, I compute the affected pixels (filter_idx), their intensities (`flat_img[filter_idx]`), and then perform a modified convolution of the form `np.min((intensities + sagitta) * kernel)` instead of `np.sum(intensities * kernel)` (or matmul(intensities, kernel)) for a standard convolution.

Does that help at all?

Reply:
Since we are discussing this part in the suggestion above I will mark this one as resolved (so that I don't lose track of what still needs to be done/has been done).